sktree/stats/permutationforest.py

```diff
@@ -0,0 +1,835 @@
+
+from .utils import METRIC_FUNCTIONS, REGRESSOR_METRICS, compute_null_distribution_perm
+
+
+class BasePermutationForest(MetaEstimatorMixin):
+    def __init__(
+        self,
+        estimator=None,
+        n_estimators=100,
+        criterion="squared_error",
+        max_depth=None,
+        min_samples_split=2,
+        min_samples_leaf=1,
+        min_weight_fraction_leaf=0.0,
+        max_features=1.0,
+        max_leaf_nodes=None,
+        min_impurity_decrease=0.0,
+        bootstrap=False,
+        oob_score=False,
+        n_jobs=None,
+        random_state=None,
+        verbose=0,
+        warm_start=False,
+        ccp_alpha=0.0,
+        max_samples=None,
+        **estimators_kwargs,
+    ):
+        self.estimator = estimator
+        self.n_jobs = n_jobs
+        self.n_estimators = n_estimators
+        self.criterion = criterion
+        self.max_depth = max_depth
+        self.min_samples_split = min_samples_split
+        self.min_samples_leaf = min_samples_leaf
+        self.min_weight_fraction_leaf = min_weight_fraction_leaf
+        self.max_features = max_features
+        self.max_leaf_nodes = max_leaf_nodes
+        self.min_impurity_decrease = min_impurity_decrease
+        self.bootstrap = bootstrap
+        self.oob_score = oob_score
+        self.random_state = random_state
+        self.verbose = verbose
+        self.warm_start = warm_start
+        self.ccp_alpha = ccp_alpha
+        self.max_samples = max_samples
+        self.estimator_kwargs = estimators_kwargs
+
+    def reset(self):
+        class_attributes = dir(type(self))
+        instance_attributes = dir(self)
+
+        for attr_name in instance_attributes:
+            if attr_name.endswith("_") and attr_name not in class_attributes:
+                delattr(self, attr_name)
+
+    def _get_estimator(self):
+        pass
+
+    def _statistic(
+        self,
+        estimator: BaseForest,
+        X: ArrayLike,
+        y: ArrayLike,
+        covariate_index: ArrayLike = None,
+        metric="mse",
+        return_posteriors: bool = False,
+        seed=None,
+        **metric_kwargs,
+    ):
+        """Helper function to compute the test statistic."""
+        metric_func = METRIC_FUNCTIONS[metric]
+        if seed is None:
+            rng = np.random.default_rng(self.random_state)
+        else:
+            rng = np.random.default_rng(seed)
+        n_samples = X.shape[0]
+        indices = np.arange(n_samples, dtype=int)
+
+        if covariate_index is not None:
+            # perform permutation of covariates
+            index_arr = rng.choice(indices, size=(X.shape[0], 1), replace=False, shuffle=False)
+            X = X.copy()
+            X[:, covariate_index] = X[index_arr, covariate_index]
+
+        X_train, X_test = X[self.indices_train_, :], X[self.indices_test_, :]
+        y_train, y_test = y[self.indices_train_, :], y[self.indices_test_, :]
```

Comment:
So this method only permutes the null individually but maintains the common training and test sets?

Reply:
Yeah I think that is the way it was done in the paper, but we can try it out w/ permuting everything as you had in `_might.py`, but just figured this should at least reproduce what was in Coleman, albeit not as "good results" it seems. 

Their simulations are a bit hard to read and interpret too